setup/helpers/puppetserver_helper.rb

```diff
@@ -0,0 +1,644 @@
+  # system
+  def strip_nils(json)
+    json.each do |json_member|
+      json_member['datapoints'].each_index do |i|
+        while ( json_member['datapoints'][i] != nil &&
+            json_member['datapoints'][i][0] == nil ) do
+          json_member['datapoints'].delete_at(i)
+        end
+      end
+    end
+  end
+
+  def metrics_target(master_hostname, metric_name)
+    return_string = "puppetlabs.#{master_hostname}.#{metric_name}"
+  end
+
+  def build_metrics_url(master, graphite_host, metrics, minutes)
+    url = "http://#{graphite_host}/render?"
+    url += "from=-#{minutes}minutes"
+    if metrics.is_a?(String)
+      url += '&target=' << metrics_target(master, metrics)
+    elsif metrics.is_a?(Array)
+      metrics.each do |m|
+        url += '&target=' << metrics_target(master, m)
+      end
+    end
+    url += '&format=json'
+  end
+
+  # returns a hash of metrics results
+  def query_metrics(master_hostname, graphite_hostname, metrics, minutes=10)
+    sleep_between_queries = 5
+    seconds_remaining = minutes * 60
+    full_results = Hash.new
+    metrics.each do |m|
+      url       = build_metrics_url(master_hostname, graphite_hostname, m, minutes)
+      uri       = URI.parse(url)
+      cont = true
+      loop do
+        response = Net::HTTP.get_response(uri)
+        assert_equal("200", response.code, 'Expected response code 200')
+        json = parse_json(response.body)
+        loop_results = strip_nils(json)
+        case
+          when (loop_results.size != 0 &&
+              loop_results[0]['datapoints'].size != 0)
+            full_results["#{master_hostname}.#{m}"] =
+                loop_results[0]['datapoints']
+            cont = false
+          when seconds_remaining > 0
+            logger.info "#{m} metric not found in graphite, waiting " \
+              "#{sleep_between_queries} seconds before trying to get it again"
+            sleep sleep_between_queries
+            seconds_remaining -= sleep_between_queries
+          else
+            cont = false
+        end
+        break if !cont
+      end
+    end
+    return full_results
+  end
+
+  # builds and returns an array of missing metrics names.
+  # expected metrics is an array of metric names, typically generated with
+  # the build_default_metrics method
+  # actual_metrics_results is an array of arrays, typically generated with the
+  # query_metrics method.
+  def build_missing_metrics_list(master_hostname, expected_metrics, actual_metrics_results)
+    missing_metrics = Array.new
+    expected_metrics.each do |m|
+      if ( !actual_metrics_results.include?("#{master_hostname}.#{m}") ||
+          !actual_metrics_results["#{master_hostname}.#{m}"].class == Array ||
+          actual_metrics_results["#{master_hostname}.#{m}"].empty? )
+        missing_metrics << "#{master_hostname}.#{m}"
+      end
+    end
+    return missing_metrics
+  end
+end
+
+Beaker::TestCase.send(:include, PuppetServerExtensions)
```

Comment:
what does this do?

Reply:
I will just do one comment for everything from puppetserver_helper.  I didn't write any of this, just copied it from pe_acceptance_tests.  We probably are not using 90% of it.  I will open a ticket to go through and remove everything we're not using, but really this is one of those cases where this should be a shared helper in it's own gem so we can reference it.

Reply:
+1 shared helper gem